otava/change_point_divisive/calculator.py

```diff
@@ -0,0 +1,189 @@
+              |                0                                  0                …  A(start, end, end + 1)       |
+
+        Matrices B and C follow the same index-to-value pattern.
+
+        Matrices A, B, and C are used to compute values of function `Q` recursively, without
+        recomputing the same sums over and over again. The recursive formulas were further
+        transformed to closed forms, so they can be computed using numpy cummulative sum
+        function to take advantage of numpy vectorized operations. (Note that each column
+        in matrices A, B, C can be repersented through np.cumsum). The formulas use
+        commulative sum of functions `H` and `V`, which definitions can be found in
+        `_calculate_distances` method.
+
+            A(start, τ, κ) = 2 / (κ - s) * (Σt=start,τ-1  H(start, t, κ) - Σt=start+1,τ-1  V(start, t, κ)),
+
+            B(start, τ, κ) = 2 * (κ - τ) / (κ - start) / (τ - start - 1) * Σt=start+1,τ-1  V(start, t, κ),
+
+            C(start, τ, κ) = 2 * (τ - start) / (κ - start) / (κ - τ - 1) * Σt=τ,κ-2  H(start, t, κ).'''
+        if self.V is None or self.H is None:
+            self._calculate_distances()
+
+        V = self.V[start : end] - self.distances[0 : start, start + 1 : end + 1].sum(axis=0)
+        H = self.H[start : end, start : end]
+
+        taus = np.arange(start + 1, end + 1)[:, None]
+        kappas = np.arange(start + 2, end + 2)[None, :]
+
+        A = np.zeros((end - start, end - start))
+        A_coefs = 2 / (kappas - start)
+        A[1:, :] = np.cumsum(V)[:-1, None]
+        A = A_coefs * np.triu(np.cumsum(H, axis=0) - A, k=0)
+
+        B = np.zeros((end - start, end - start))
+        B_num = 2 * (kappas - taus)
+        B_den = (kappas - start) * (taus - start - 1)
+        B_mask = np.triu(np.ones_like(B_den, dtype=bool), k=0)
+        B_out = np.zeros_like(B_den, dtype=float)
+        B_coefs = np.divide(B_num, B_den, out=B_out, where=B_mask & (B_den != 0))
+        B[1:, 1:] = B_coefs[1:, 1:] * np.cumsum(V)[:-1, None]
+
+        C = np.zeros((end - start, end - start))
+        C_num = 2 * (taus - start)
+        C_den = (kappas - start) * (kappas - taus - 1)
+        C_mask = np.triu(np.ones_like(C_den, dtype=bool), k=1)
+        C_out = np.zeros_like(C_den, dtype=float)
+        C_coefs = np.divide(C_num, C_den, out=C_out, where=C_mask & (C_den != 0))
+        C[:-1, 1:] = C_coefs[:-1, 1:] * np.flipud(np.cumsum(np.flipud(H[1:, 1:]), axis=0))
+
+        # Element of matrix `Q_{i, j}` is equal to `Q(τ, κ) = Q(i + 1, j + 2) = QQ(sequence[start : i + 1], sequence[i + 1 : j + 2])`.
+        # So, critical point is `τ = i + 1`.
+        return A - B - C
+
+    def get_candidate_change_point(self, interval: slice) -> CandidateChangePoint:
+        '''For a given `slice(start, stop)` finds potential critical point in subsequence series[slice],
+        i.e., from index `start` to `stop - 1` inclusive. For simplicity, we'll use `end = stop - 1`.
```

Comment:
To be super clear, maybe use "interval" and ,) if you are talking math, and slice with [] and range with () in python context